cola/ace_generator.py

```diff
@@ -0,0 +1,90 @@
+
+class COLA:
+    def __init__(self, ml_model=None, data=None, counterfactual_algorithm=None, policy_name="CF_OTMatch", Avalues_method="max", limited_actions=None):
+        """
+        Initialize the ACE_generator class
+
+        Parameters:
+        ml_model: Machine learning model used for predictions
+        data: DataFrame containing the data
+        counterfactual_algorithm: Algorithm to generate counterfactual data
+        limited_actions: List of limited actions to consider
+        """
+        self.ml_model = ml_model
+        self.data = data
+        self.counterfactual_algorithm = counterfactual_algorithm
+        self.limited_actions = limited_actions
+        self.Avalues_method = Avalues_method
+        self.policy_name = policy_name
+        # self.ce_algorithm =None
+        # self.results = self.generate_counterfactual()
+
+    def generate_ace_counterfactual(self):
+        
+        # get model
+        self.ml_model_true = self.ml_model.load_model()
+
+        # get ce_algorithm
+        ce_algorithm = self.counterfactualexplanation_algorithm()
+        x_factual, y_factual, x_counterfactual, y_counterfactual = ce_algorithm.generate_x_counterfactuals()
+
+        # get policy
+        policies = policy.compute_intervention_policy(
+                            model= self.ml_model_true,   
+                            X_train=0,    # X_train=x_train didn't use the X_train in this part
+                            X_factual=x_factual,
+                            X_counterfactual=x_counterfactual,
+                            shapley_method=self.policy_name,
+                            Avalues_method=self.Avalues_method, # 'avg'
+                        )
+        varphi = policies['varphi']
+        p = policies['p']
+        q = policies['q']
+
+        # 1. Find the top 'action' highest probability values and their positions in the varphi matrix
+        flat_indices = np.argpartition(varphi.flatten(), -self.limited_actions)[-self.limited_actions:]
+        row_indices, col_indices = np.unravel_index(flat_indices, varphi.shape)
+
+        # 2. Find the values at these positions in q
+        q_values = q[row_indices, col_indices]
+
+        # 3. Replace the corresponding values in x_factual with the values found in q
+        x_action_constrained = x_factual.copy()
+
+        # 4. get the action-constrained CE
+        for row_idx, col_idx, q_val in zip(row_indices, col_indices, q_values):
+            x_action_constrained[row_idx, col_idx] = q_val
+        
+        # 5. get the prediction of action-constrained CE
+        y_counterfactual = self.ml_model_true.predict(x_action_constrained)
+
+        factual = self.return_dataframe(x_factual, y_factual)
+        ce = self.return_dataframe(x_counterfactual, y_counterfactual)
+        ace = self.return_dataframe(x_action_constrained, y_counterfactual)
+        
+        return factual, ce, ace
+
+
+    def counterfactualexplanation_algorithm(self):
```

Comment:
Later on, you gonna support multiple counterfactual explanations algorithms. I think it is better to explicitly declare as:

def counterfactual_explanation_algorithm(self) -> CounterFactualExplainer: